app/app.go

```diff
@@ -168,8 +174,52 @@ func (app *App) InitChain(req abci.RequestInitChain) (res abci.ResponseInitChain
 	return
 }
 
-// Implements ABCI
+// Query implements ABCI and allows use to externally inspect state
 func (app *App) Query(req abci.RequestQuery) (res abci.ResponseQuery) {
+	// TODO: actually implement this
+	if req.Prove {
+		return abci.ResponseQuery{Code: 500, Log: "Query with proof not implemented"}
+	}
+	if req.Height != 0 {
+		return abci.ResponseQuery{Code: 500, Log: "Historical query not implemented"}
+	}
+
+	// First step is to figure which substore to route the query to....
+	path := req.Path
+	switch path {
+	// To keep backwards compatibility ???
```

Comment:
Lets not worry about backwards compatibility.  We don't have to support all the fields in abci.RequestQuery or abci.ResponseQuery either.

The design constraint that I strongly support for Query is for it to be feature compatible with HTTP queries.  So that means, the RequestQuery.Path should be the URL path, and it may or may not include the parameters... we should probably make ReqeustQuery.  Currently it is only:

```proto
message RequestQuery {
  bytes data = 1;
  string path = 2;
  int64 height = 3;
  bool prove = 4;
}
```

How about we evolve it toward:

```proto
message RequestQuery {
  bytes data = 1; // in case Query came as a POST, and the query includes bytes.
  string path = 2; // e.g. the URL path
  // int64 height = 3; DEPRECATED
  // bool prove = 4; DEPRECATED
  repeated common.KVPair params = 5; // "h"/"height" and "p"/"prove" are params here.
  repeated accept string = 6; // Similar to HTTP "Accept" header, e.g. "json" or "wire".  "json/unpack" or "wire/unpack" could mean, use the specific ResponseQuery fields, otherwise the response bytes can just be in the ResponseQuery.value field as bytes.
}
```

Reply:
I've added an issue here: https://github.com/tendermint/abci/issues/182

Reply:
I do not like this.

We should pick an internal format that is clearly parsed, with clear semantics. What we currently have with the object. This can be converted to/from JSON/URI/protobuf/HTML/etc. We do not want to worry about that in the abci interface. That is trying all the parts too tightly together.

The current abci QueryRequest is basically fine. I suggest we just add the parsing code in the RPC server to convert that into a QueryRequest object. If there are new fields we need, we should add them.

But generic `repeated common.KVPair` is even worse than the `type Values map[string][]string` we get from `net/http`